### PR TITLE
Exchanges: map rendering

### PR DESCRIPTION
**Pull request by Kongkille**
## Issue
https://linear.app/electricitymaps/issue/ELE-1369/set-up-exchanges-layer

## Description

This was a bit complicated. I tried to copy as much as possible but a few things seem to have been deprecated

- BaseControl is no longer accessible, so I had to create custom layers in another way
- OnViewPortChange also seems deprecated so we need to monitor isMoving on direct handlers such as onDrag end start etc. This is actually pretty nice, so we don't have the custom debounce logic.
- Easing has been added so we need to turn it off with DragPan settings
- A few other things that I had to change due to either me not understanding it, or the libraries having changed functionality somehow.

Performance is... ok. Similar to the current version. I had hoped that I could easily tweak some things and get good performance but the current approach just 

I think part of it comes directly from the useExchangeArrows hook, so we could optimize that otherwise I think we need to go back to the whiteboard to get really good performance.

### Preview

https://user-images.githubusercontent.com/45588799/202518823-474194e6-dcf9-4b13-afd7-1928f468feb9.mov

### CODE REVIEW

**Review by tonypls — COMMENTED** [review]
LGTM

**Comment by Kongkille** [comment]
I can do the arrows in another PR if it's annoying to look at

**Review by tonypls — COMMENTED** [review]
Works well, locally. 

I find the custom layer wrapper stuff a bit confusing, not your code but the way the library is meant to be used. Would be great if we could go through the PR tomorrow so I can understand it a bit better and provide some more constructive feedback.

File: web/src/features/map/Map.tsx [hunk 1]
```diff
@@ -198,6 +201,23 @@ export default function MapPage(): ReactElement {
...
+    setIsMoving(false);
+  };
+
+  const onZoomStart = () => {
+    setIsMoving(true);
+  };
+
+  const onZoomEnd = () => {
+    setIsMoving(false);
+  };
```
Comment (VIKTORVAV99) [inline web/src/features/map/Map.tsx:219]: While I realize why you did it this way and it's super clear I don't think we should duplicate exact functionality. Perhaps combining them like the following suggestion would work?
 ```suggestion
  const onDragOrZoomStart = () => {
    setIsMoving(true);
  };

  const onDragOrZoomEnd = () => {
    setIsMoving(false);
  };
```
Naming is just a suggestion, feel free to change it to something else if you want to go forward with this.

File: web/src/features/map/Map.tsx [hunk 2]
```diff
@@ -214,6 +234,11 @@ export default function MapPage(): ReactElement {
         onError={onError}
         onMouseMove={onMouseMove}
         onMouseOut={onMouseOut}
+        onDragStart={onDragStart}
+        onZoomStart={onZoomStart}
+        onZoomEnd={onZoomEnd}
+        dragPan={{ maxSpeed: 0 }} // Disables easing effect to improve performance on exchange layer
+        onDragEnd={onDragEnd}
```
Comment (VIKTORVAV99) [inline web/src/features/map/Map.tsx:241]: Follow up suggestion to the first one:
```suggestion
        onDragStart={onDragOrZoomStart}
        onZoomStart={onDragOrZoomStart}
        onZoomEnd={onDragOrZoomEnd}
        dragPan={{ maxSpeed: 0 }} // Disables easing effect to improve performance on exchange layer
        onDragEnd={onDragOrZoomEnd}
```

File: web/src/features/exchanges/ExchangeArrow.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  data: ExchangeArrowData;
+  viewportWidth: number;
+  viewportHeight: number;
+  map: MapboxMap;
+}
+
+function ExchangeArrow({ data, viewportWidth, viewportHeight, map }: ExchangeArrowProps) {
+  const mapZoom = map.getZoom();
+  const colorBlindModeEnabled = false; // TODO: FIX
+  const absFlow = Math.abs(data.netFlow || 0);
```
Comment (VIKTORVAV99) [inline web/src/features/exchanges/ExchangeArrow.tsx:18]: Should this use nullish coalescing instead?
```suggestion
  const absFlow = Math.abs(data.netFlow ?? 0);
```

File: web/src/utils/scales.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+//   .domain([0, 500, 1000])
+//   .range(['black', 'transparent', 'gold'])
+//   .clamp(true);
+
+// ** Exchange
+
+export const quantizedCo2IntensityScale = scaleQuantize()
+  .domain([0, 800])
+  .range([0, 80, 160, 240, 320, 400, 480, 560, 640, 720, 800])
+  .unknown('nan');
```
Comment (VIKTORVAV99) [inline web/src/utils/scales.ts:33]: Not really related to this PR in particular but we should probably bump the range from 0-800 to 0-1500 as discussed in this issue: https://github.com/electricitymaps/electricitymaps-contrib/issues/4771

**Review by VIKTORVAV99 — COMMENTED** [review]
Just some thoughts and questions. 🙂

File: web/src/features/exchanges/ExchangeArrow.tsx (hunk 3 above)
Comment (Kongkille) [reply web/src/features/exchanges/ExchangeArrow.tsx:18]: Yes :)

File: web/src/utils/scales.ts (hunk 4 above)
Comment (Kongkille) [reply web/src/utils/scales.ts:33]: I'll leave the for a future change for now :)

File: web/src/features/map/Map.tsx (hunk 1 above)
Comment (Kongkille) [reply web/src/features/map/Map.tsx:219]: Sure

File: web/src/features/exchanges/ExchangeArrow.tsx [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    return null;
+  }
+
+  if (transform.y - 100 * transform.k > viewportHeight) {
+    return null;
+  }
+
+  return (
+    <div
+      id={key}
```
[the remark this answers is not in the document]
Comment (Kongkille) [reply web/src/features/exchanges/ExchangeArrow.tsx:70]: Yeah I tried, it didn't work. Not sure how to get that working.

It's the same as in prod atm.

**Comment by Kongkille** [comment]
> Works well, locally.
> 
> I find the custom layer wrapper stuff a bit confusing, not your code but the way the library is meant to be used. Would be great if we could go through the PR tomorrow so I can understand it a bit better and provide some more constructive feedback.

To be honest, I don't think I fully understand it either but we can try :)

**Review by tonypls — APPROVED** [review]
👍

**Pull request merged**